File: lexer/lexer.go
// Package lexer: contains the lexer
package lexer

import (
	"mycelium/token"
)

type Lexer struct {
	input        string
	position     int  // current position in input
	readPosition int  // points to the h after position
	ch           byte // current char
}

func New(input string) *Lexer {
	lexer := &Lexer{input: input}
	lexer.readChar()
	return lexer
}

func (lex *Lexer) NextToken() token.Token {
	var tok token.Token
	lex.skipWhitespace()
	switch lex.ch {
	case '=':
		if isComparator(lex.ch) {
			tok.Literal = lex.readComparator()
			if len(tok.Literal) == 2 {
				tok.Type = token.LookUpComparator(tok.Literal)
				return tok
			}
			tok.Type = token.ASSIGN
		}
		return tok
	case '+':
		tok = newToken(token.PLUS, lex.ch)
	case '-':
		tok = newToken(token.MINUS, lex.ch)
	case '*':
		tok = newToken(token.MULT, lex.ch)
	case '/':
		tok = newToken(token.DIV, lex.ch)
	case '\\':
		tok = newToken(token.BACKSLASH, lex.ch)
	case '<':
		if isComparator(lex.ch) {
			tok.Literal = lex.readComparator()
			if len(tok.Literal) == 2 {
				tok.Type = token.LookUpComparator(tok.Literal)
				return tok
			}
			tok.Type = token.ASSIGN
		}
		return tok
	case '>':
		if isComparator(lex.ch) {
			tok.Literal = lex.readComparator()
			if len(tok.Literal) == 2 {
				tok.Type = token.LookUpComparator(tok.Literal)
				return tok
			}
			tok.Type = token.ASSIGN
		}
		return tok
	case '!':
		tok = newToken(token.BANG, lex.ch)
	case ',':
		tok = newToken(token.COMMA, lex.ch)
	case ';':
		tok = newToken(token.SEMICOLON, lex.ch)
	case '(':
		tok = newToken(token.OBRACKET, lex.ch)
	case ')':
		tok = newToken(token.CBRACKET, lex.ch)
	case '{':
		tok = newToken(token.OCURLBRACKET, lex.ch)
	case '}':
		tok = newToken(token.CCURLBRACKET, lex.ch)
	case '[':
		tok = newToken(token.OCORNBRACKET, lex.ch)
	case ']':
		tok = newToken(token.CCORNBRACKE, lex.ch)
	case ':':
		tok = newToken(token.COLON, lex.ch)
	case '"':
		tok.Literal = lex.readString()
		tok.Type = token.STRING
	case 0:
		tok.Literal = ""
		tok.Type = token.EOF
	default:
		if isLetter(lex.ch) {
			tok.Literal = lex.readIdentifier()
			tok.Type = token.LookupIdent(tok.Literal)
			return tok
		} else if isDigit(lex.ch) {
			tok.Type = token.INT
			tok.Literal = lex.readNumber()
			return tok
		} else {
			tok = newToken(token.ILLEGAL, lex.ch)
		}
	}

	lex.readChar()
	return tok
}

func newToken(tokenType token.TokenType, ch byte) token.Token {
	return token.Token{Type: tokenType, Literal: string(ch)}
}

func (lex *Lexer) skipWhitespace() {
	for lex.ch == ' ' || lex.ch == '\t' || lex.ch == '\n' || lex.ch == '\r' {
		lex.readChar()
	}
}

func (lex *Lexer) readString() string {
	lex.readChar()
	startposition := lex.position
	var endposition int
	for {
		if lex.ch == '"' || lex.ch == 0 {
			endposition = lex.position
			break
		}
		lex.readChar()
	}
	return lex.input[startposition:endposition]
}

func (lex *Lexer) readIdentifier() string {
	position := lex.position
	for isLetter(lex.ch) {
		lex.readChar()
	}
	return lex.input[position:lex.position]
}

func (lex *Lexer) readComparator() string {
	position := lex.position
	for isComparator(lex.ch) {
		lex.readChar()
	}
	return lex.input[position:lex.position]
}

func (lex *Lexer) readChar() {
	if lex.readPosition >= len(lex.input) {
		lex.ch = 0
	} else {
		lex.ch = lex.input[lex.readPosition]
	}
	lex.position = lex.readPosition
	lex.readPosition += 1
}

func (lex *Lexer) readNumber() string {
	position := lex.position
	for isDigit(lex.ch) {
		lex.readChar()
	}
	return lex.input[position:lex.position]
}

func isDigit(ch byte) bool {
	return '0' <= ch && ch <= '9'
}

func isComparator(ch byte) bool {
	return '<' <= ch && ch <= '>'
}

func isLetter(ch byte) bool {
	return 'A' <= ch && ch <= 'Z' || 'a' <= ch && ch <= 'z' || ch == '_'
}

File: src/lexer/lexer.rs
use crate::token::token::Token;

pub struct Lexer {
    pub input: String,
    pub position: usize,
    pub next_position: usize,
    pub char: u8,
}

impl Lexer {
    pub fn new(input: String) -> Lexer {
        let mut lexer = Lexer {
            input,
            position: 0,
            next_position: 0,
            char: 0,
        };
        lexer.read_char();
        lexer
    }

    pub(crate) fn next_token(&mut self) -> Token {
        self.skip_whitespace();
        let token = match self.char {
            b'=' => {
                if self.peek_char() == b'=' {
                    self.read_char();
                    Token::EQ
                } else {
                    Token::ASSIGN
                }
            }
            b'+' => Token::PLUS,
            b'-' => Token::MINUS,
            b'*' => Token::MULTIPLY,
            b'/' => Token::DIVIDE,
            b'\\' => Token::BACKSLASH,
            b'<' => {
                if self.peek_char() == b'=' {
                    self.read_char();
                    Token::LESSERTHENEQ
                } else {
                    Token::LESSTHAN
                }
            }
            b'>' => {
                if self.peek_char() == b'=' {
                    self.read_char();
                    Token::GREATERTHANEQ
                } else {
                    Token::GREATERTHAN
                }
            }
            b'!' => {
                if self.peek_char() == b'=' {
                    self.read_char();
                    Token::NOTEQUAL
                } else {
                    Token::BANG
                }
            }
            b',' => Token::COMMA,
            b';' => Token::SEMICOLON,
            b'(' => Token::OPENBRACKET,
            b')' => Token::CLOSEBRACKET,
            b'{' => Token::OPENCURLBRACKET,
            b'}' => Token::CLOSEDCURLBRACKET,
            b'[' => Token::OPENCORNERBRACKET,
            b']' => Token::CLOSEDCORNERBRACKET,
            b':' => Token::COLON,
            b'"' => {
                let string = self.read_string();
                Token::STRING(string)
            }
            0 => Token::EOF("EOF".to_string()),
            _ => {
                return if is_letter(self.char) {
                    let identifier = self.read_identifier();
                    Token::lookup_identifier(&identifier)
                } else if self.char.is_ascii_digit() {
                    let number = self.read_number();
                    Token::INT(number)
                } else {
                    let token = Token::ILLEGAL(String::from_utf8_lossy(&[self.char]).to_string());
                    self.read_char();
                    token
                };
            }
        };
        self.read_char();
        token
    }

    fn skip_whitespace(&mut self) {
        while self.char.is_ascii_whitespace() {
            self.read_char();
        }
    }

    fn read_string(&mut self) -> String {
        self.read_char();
        let start_position = self.position;
        let end_position: usize;

        loop {
            if self.char == u8::try_from('"').unwrap() || self.char == 0 {
                end_position = self.position;
                break;
            }
            self.read_char();
        }

        self.input[start_position..end_position].to_string()
    }

    fn read_identifier(&mut self) -> String {
        let position = self.position;
        while is_letter(self.char) {
            self.read_char();
        }
        self.input[position..self.position].to_string()
    }

    fn read_number(&mut self) -> String {
        let position = self.position;
        while self.char.is_ascii_digit() {
            self.read_char();
        }
        self.input[position..self.position].to_string()
    }

    fn read_char(&mut self) {
        if self.next_position >= self.input.len() {
            self.char = 0;
        } else {
            self.char = self.input.as_bytes()[self.next_position];
        }
        self.position = self.next_position;
        self.next_position += 1;
    }

    fn peek_char(&self) -> u8 {
        if self.next_position >= self.input.len() {
            0
        } else {
            self.input.as_bytes()[self.next_position]
        }
    }
}

fn is_letter(ch: u8) -> bool {
    ch.is_ascii_alphabetic() || ch == b'_'
}
